File: apps/macos-menubar/Sources/MurmurTrayCore/Updates.swift
import Foundation

public enum UpdateError: Error, LocalizedError, Sendable {
    case invalidResponse, invalidReleasePage
    public var errorDescription: String? {
        switch self {
        case .invalidResponse: L10n.text("CLI returned an unverified update result")
        case .invalidReleasePage: L10n.text("CLI did not confirm the official release page")
        }
    }
}

/// A consumer of the shared CLI result. No version comparison, HTTP or cache paths.
public struct UpdateSnapshot: Decodable, Sendable {
    public enum State: String, Decodable, Sendable { case current = "up-to-date", available, unknown }
    public let schema: String, channel: String, versionSource: String, comparison: String
    public let currentVersion: String?, latestVersion: String?, releaseUrl: String?, action: String?
    public let enabled: Bool, state: State, reason: String
    public let checkedAt: String?, lastSuccessAt: String?, nextCheckAt: String?
    public let cached: Bool, stale: Bool
    public let checkIntervalMs: Int, timeoutMs: Int

    public static func decode(_ data: Data) throws -> Self {
        let required: Set<String> = ["schema", "channel", "currentVersion", "versionSource", "comparison", "enabled",
            "state", "reason", "latestVersion", "releaseUrl", "action", "checkedAt", "lastSuccessAt", "nextCheckAt",
            "cached", "stale", "checkIntervalMs", "timeoutMs"]
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              required.isSubset(of: Set(object.keys)),
              let value = try? JSONDecoder().decode(Self.self, from: data),
              schemaKnown(value.schema, name: "murmur.updates"), value.channel == "stable",
              value.versionSource == "root-package-json", value.comparison == "declared-release-version",
              value.checkIntervalMs == 21_600_000, value.timeoutMs == 4_000,
              value.reason.range(of: "\\Aupdates\\.[a-z0-9.-]{1,100}\\z", options: .regularExpression) != nil else {
            throw UpdateError.invalidResponse
        }
        for version in [value.currentVersion, value.latestVersion].compactMap({ $0 }) {
            guard !version.isEmpty, version.count <= 128,
                  !version.unicodeScalars.contains(where: { CharacterSet.controlCharacters.contains($0) }) else {
                throw UpdateError.invalidResponse
            }
        }
        for date in [value.checkedAt, value.lastSuccessAt, value.nextCheckAt].compactMap({ $0 }) {
            guard timestamp(date) != nil else { throw UpdateError.invalidResponse }
        }
        if value.state == .unknown {
            guard value.action == nil, value.releaseUrl == nil else { throw UpdateError.invalidResponse }
        } else {
            guard value.enabled, !value.stale, value.currentVersion != nil, value.latestVersion != nil,
                  value.checkedAt != nil, value.lastSuccessAt != nil else { throw UpdateError.invalidResponse }
            if value.state == .available {
                guard value.reason == "updates.newer-release", value.action == "open-release-page",
                      let page = value.releaseUrl, officialReleasePage(page) != nil else { throw UpdateError.invalidReleasePage }
            } else {
                guard value.reason == "updates.no-newer-release", value.action == nil,
                      value.releaseUrl == nil else { throw UpdateError.invalidResponse }
            }
        }
        return value
    }

    public static func officialReleasePage(_ value: String) -> URL? {
        let prefix = "/alexfrmn/murmur/releases/tag/"
        guard let parts = URLComponents(string: value), parts.scheme == "https", parts.host == "github.com",
              parts.user == nil, parts.password == nil, parts.port == nil, parts.query == nil, parts.fragment == nil,
              parts.percentEncodedPath.hasPrefix(prefix) else { return nil }
        guard let tag = String(parts.percentEncodedPath.dropFirst(prefix.count)).removingPercentEncoding,
              !tag.contains(".."), tag.range(of: "\\A[A-Za-z0-9][A-Za-z0-9._+-]{0,127}\\z", options: .regularExpression) != nil else { return nil }
        return parts.url
    }

    public func expired(now: Date = Date()) -> Bool {
        guard let checkedAt, let checked = timestamp(checkedAt) else { return true }
        return now.timeIntervalSince(checked) > Double(checkIntervalMs) / 1_000 || checked.timeIntervalSince(now) > 5
    }
    public func releasePage(now: Date = Date()) -> URL? {
        guard state == .available, enabled, !stale, !expired(now: now), action == "open-release-page",
              let releaseUrl else { return nil }
        return Self.officialReleasePage(releaseUrl)
    }
    public func title(now: Date = Date()) -> String {
        if !enabled && reason == "updates.disabled" { return L10n.text("Update checks are disabled") }
        if state != .unknown && expired(now: now) { return L10n.text("The update result is stale") }
        switch state {
        case .available: return L10n.text("Version %@ is available", String(describing: (latestVersion ?? L10n.text("unknown"))))
        case .current: return L10n.text("No newer stable release was found")
        case .unknown: return L10n.text("Updates could not be checked")
        }
    }
    public var reasonText: String {
        switch reason {
        case "updates.disabled": L10n.text("Automatic checks are disabled")
        case "updates.current-version-invalid": L10n.text("CLI could not determine the product version")
        case "updates.preferences-unavailable": L10n.text("The update preference is unavailable")
        case "updates.cache-invalid": L10n.text("The saved result is invalid")
        case "updates.cache-unavailable": L10n.text("The saved result is unavailable")
        case "updates.check-in-progress": L10n.text("Another process is already checking")
        case "updates.interrupted": L10n.text("The previous check was interrupted")
        case "updates.network-error": L10n.text("Could not reach GitHub")
        case "updates.timeout": L10n.text("GitHub did not respond in time")
        case "updates.rate-limited": L10n.text("GitHub limited the request rate")
        case "updates.http-error": L10n.text("GitHub returned an error")
        case "updates.release-invalid": L10n.text("GitHub did not confirm a stable release")
        case "updates.newer-release": L10n.text("A newer stable release is available")
        case "updates.no-newer-release": L10n.text("Compared using the declared release version")
        default: L10n.text("CLI reported: %@", String(describing: (reason)))
        }
    }
    public func ageText(now: Date = Date()) -> String {
        guard let checkedAt, let date = timestamp(checkedAt) else { return L10n.text("No network check has been recorded") }
        let age = max(0, Int(now.timeIntervalSince(date) / 60))
        return L10n.text("%@ · %@ minutes ago", String(describing: (cached ? L10n.text("Cached") : L10n.text("Checked"))), String(describing: (age)))
    }
}

public struct UpdatesClient: Sendable {
    public let executable: URL
    public let forcedOff: Bool
    private let environment: [String: String]
    private let timeout: TimeInterval
    public init(executable: URL, environment: [String: String] = ProcessInfo.processInfo.environment,
                timeout: TimeInterval = 6) {
        self.executable = executable; self.environment = environment; self.timeout = timeout
        forcedOff = environment["MURMUR_UPDATE_CHECK"] == "0"
    }
    public func check() throws -> UpdateSnapshot {
        let result = try CLIProbe(executable: executable, timeout: timeout, environment: environment).invoke(["updates", "check"])
        return try UpdateSnapshot.decode(result.data)
    }
    public func setEnabled(_ enabled: Bool) throws {
        let result = try CLIProbe(executable: executable, timeout: timeout, environment: environment)
            .invoke(["updates", enabled ? "enable" : "disable"])
        struct Preference: Decodable { let schema: String, enabled: Bool }
        guard let value = try? JSONDecoder().decode(Preference.self, from: result.data),
              schemaKnown(value.schema, name: "murmur.update-preferences"), value.enabled == enabled else {
            throw UpdateError.invalidResponse
        }
    }
}

File: spikes/windows-tray-go/updates.go
package main

// Consume the shared CLI contract. Version comparison, networking and cache
// selection belong to the engine, never to the tray.
import (
	"context"
	"encoding/json"
	"errors"
	"net/url"
	"regexp"
	"strings"
	"time"
	"unicode"
)

const updateInterval = 6 * time.Hour

var errUpdateResponse = errors.New("update response was not confirmed")
var releaseTag = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._+-]{0,127}$`)
var updateReason = regexp.MustCompile(`^updates\.[a-z0-9.-]{1,100}$`)

type updateSnapshot struct {
	Schema          string  `json:"schema"`
	Channel         string  `json:"channel"`
	CurrentVersion  *string `json:"currentVersion"`
	VersionSource   string  `json:"versionSource"`
	Comparison      string  `json:"comparison"`
	Enabled         bool    `json:"enabled"`
	State           string  `json:"state"`
	Reason          string  `json:"reason"`
	LatestVersion   *string `json:"latestVersion"`
	ReleaseURL      *string `json:"releaseUrl"`
	Action          *string `json:"action"`
	CheckedAt       *string `json:"checkedAt"`
	LastSuccessAt   *string `json:"lastSuccessAt"`
	NextCheckAt     *string `json:"nextCheckAt"`
	Cached          bool    `json:"cached"`
	Stale           bool    `json:"stale"`
	CheckIntervalMs int     `json:"checkIntervalMs"`
	TimeoutMs       int     `json:"timeoutMs"`
}

func officialReleasePage(raw string) bool {
	u, err := url.Parse(raw)
	const prefix = "/alexfrmn/murmur/releases/tag/"
	if err != nil || u.Scheme != "https" || u.Host != "github.com" || u.User != nil || u.RawQuery != "" || u.ForceQuery || u.Fragment != "" || strings.Contains(raw, "#") || !strings.HasPrefix(u.EscapedPath(), prefix) {
		return false
	}
	tag, err := url.PathUnescape(strings.TrimPrefix(u.EscapedPath(), prefix))
	return err == nil && !strings.Contains(tag, "..") && releaseTag.MatchString(tag)
}

func decodeUpdates(data []byte) (*updateSnapshot, error) {
	var fields map[string]json.RawMessage
	var value updateSnapshot
	if json.Unmarshal(data, &fields) != nil || json.Unmarshal(data, &value) != nil {
		return nil, errUpdateResponse
	}
	nullable := map[string]bool{"currentVersion": true, "latestVersion": true, "releaseUrl": true, "action": true, "checkedAt": true, "lastSuccessAt": true, "nextCheckAt": true}
	for _, key := range []string{"schema", "channel", "currentVersion", "versionSource", "comparison", "enabled", "state", "reason", "latestVersion", "releaseUrl", "action", "checkedAt", "lastSuccessAt", "nextCheckAt", "cached", "stale", "checkIntervalMs", "timeoutMs"} {
		raw, ok := fields[key]
		if !ok || (!nullable[key] && string(raw) == "null") {
			return nil, errUpdateResponse
		}
	}
	if !schemaKnown(value.Schema, "murmur.updates/1") || value.Channel != "stable" || value.VersionSource != "root-package-json" || value.Comparison != "declared-release-version" || value.CheckIntervalMs != int(updateInterval/time.Millisecond) || value.TimeoutMs != 4000 || !updateReason.MatchString(value.Reason) {
		return nil, errUpdateResponse
	}
	for _, version := range []*string{value.CurrentVersion, value.LatestVersion} {
		if version != nil && (len(*version) == 0 || len(*version) > 128 || strings.IndexFunc(*version, unicode.IsControl) >= 0) {
			return nil, errUpdateResponse
		}
	}
	for _, stamp := range []*string{value.CheckedAt, value.LastSuccessAt, value.NextCheckAt} {
		if stamp != nil {
			if _, err := time.Parse(time.RFC3339Nano, *stamp); err != nil {
				return nil, errUpdateResponse
			}
		}
	}
	switch value.State {
	case "unknown":
		if value.Action != nil || value.ReleaseURL != nil || value.Reason == "updates.newer-release" || value.Reason == "updates.no-newer-release" {
			return nil, errUpdateResponse
		}
	case "available", "up-to-date":
		if !value.Enabled || value.Stale || value.CurrentVersion == nil || value.LatestVersion == nil || value.CheckedAt == nil || value.LastSuccessAt == nil {
			return nil, errUpdateResponse
		}
		if value.State == "available" {
			if value.Reason != "updates.newer-release" || value.Action == nil || *value.Action != "open-release-page" || value.ReleaseURL == nil || !officialReleasePage(*value.ReleaseURL) {
				return nil, errUpdateResponse
			}
		} else if value.Reason != "updates.no-newer-release" || value.Action != nil || value.ReleaseURL != nil {
			return nil, errUpdateResponse
		}
	default:
		return nil, errUpdateResponse
	}
	return &value, nil
}

func (s *updateSnapshot) expired(now time.Time) bool {
	if s == nil || s.CheckedAt == nil {
		return true
	}
	at, err := time.Parse(time.RFC3339Nano, *s.CheckedAt)
	return err != nil || now.Sub(at) > updateInterval || at.Sub(now) > 5*time.Second
}
func (s *updateSnapshot) page(now time.Time) string {
	if s == nil || s.State != "available" || !s.Enabled || s.Stale || s.expired(now) || s.Action == nil || *s.Action != "open-release-page" || s.ReleaseURL == nil || !officialReleasePage(*s.ReleaseURL) {
		return ""
	}
	return *s.ReleaseURL
}
func (s *updateSnapshot) title(now time.Time) string {
	if s == nil {
		return tr("updates.notChecked")
	}
	if !s.Enabled && s.Reason == "updates.disabled" {
		return tr("updates.disabled")
	}
	if s.State != "unknown" && s.expired(now) {
		return tr("updates.expired")
	}
	switch s.State {
	case "available":
		return tr("updates.available", *s.LatestVersion)
	case "up-to-date":
		return tr("updates.none")
	default:
		return tr("updates.failed")
	}
}
func (s *updateSnapshot) observation(now time.Time) string {
	if s == nil || s.CheckedAt == nil {
		return tr("updates.noCheck")
	}
	at, _ := time.Parse(time.RFC3339Nano, *s.CheckedAt)
	age := int(now.Sub(at) / time.Minute)
	if age < 0 {
		age = 0
	}
	label := tr("updates.checked")
	if s.Cached {
		label = tr("updates.cached")
	}
	return tr("updates.observation", label, at.UTC().Format("2006-01-02 15:04:05 UTC"), age)
}

func updateReasonLabel(reason string) string {
	if reason == "" {
		return ""
	}
	key := "updates.reason." + strings.TrimPrefix(reason, "updates.")
	if value, ok := catalogs[currentLocale()][key]; ok {
		return value
	}
	return reason
}

// Manual checks preserve consent and use the same CLI cache as scheduled checks.
// Each subprocess keeps its own existing timeout, including preference changes.
func performUpdateRequest(parent context.Context, preference *bool, limit time.Duration) (*updateSnapshot, error) {
	if preference != nil {
		ctx, cancel := context.WithTimeout(parent, limit)
		err := setUpdatesEnabled(ctx, *preference)
		cancel()
		if err != nil {
			return nil, err
		}
	}
	ctx, cancel := context.WithTimeout(parent, limit)
	defer cancel()
	return fetchUpdates(ctx)
}

func fetchUpdates(ctx context.Context) (*updateSnapshot, error) {
	data, err := runRaw(ctx, "updates", "check", "--json")
	if err != nil {
		return nil, err
	}
	return decodeUpdates(data)
}
func setUpdatesEnabled(ctx context.Context, enabled bool) error {
	action := "disable"
	if enabled {
		action = "enable"
	}
	data, err := runRaw(ctx, "updates", action, "--json")
	if err != nil {
		return err
	}
	var value struct {
		Schema  string `json:"schema"`
		Enabled *bool  `json:"enabled"`
	}
	if json.Unmarshal(data, &value) != nil || !schemaKnown(value.Schema, "murmur.update-preferences/1") || value.Enabled == nil || *value.Enabled != enabled {
		return errUpdateResponse
	}
	return nil
}
